src/main/java/org/mariadb/jdbc/internal/failover/AbstractMastersListener.java

```diff
@@ -352,7 +416,28 @@ public FailLoop(Listener listener) {
             this.listener = listener;
         }
 
-        public void run() {
+        public void scheduleSelf(ScheduledExecutorService scheduler, boolean now) {
+            scheduledFuture = scheduler.scheduleWithFixedDelay(this, now ? 0 : 250, 250, 
+                                                               TimeUnit.MILLISECONDS);
+        }
+
+        // while scheduledFuture may be set from another thread
+        // it is expected only one thread will unschedule
+        @Override
+        public void unscheduleTask() {
+            while (true) {
+                if (scheduledFuture == null) {
+                    // not scheduled yet, should be a rare condition, just loop
+                    Thread.yield();
+                } else {
+                    scheduledFuture.cancel(false);
+                    scheduledFuture = null;
```

Comment:
I just pushed a fix which includes the return.  I also just made the AtomicReference private, since it seems like we are never invoking "blockTillTerminated" outside of stopFailover anyways.  So we are already guarding that behavior and wont have the above condition.  We just need to be sure that attempts to stop the fail loop always go through "stopFailloop", and not to the loop directly.

I will say, it's a bit confusing (I was confused when I made the above comment), it probably would be worth doing my second bullet at some point to move more of the FailLoop thread safety into FailLoop, and just have a single reference which we invoke to schedule and unschedule itself.  Just a thought.

Thanks for the feedback!  what a silly mistake on my part.  Let me know if you have any other issues or questions.